cpp/mg_utility/mg_utils.hpp

```diff
@@ -101,34 +105,55 @@ class OnScopeExit {
   std::function<void()> function_;
 };
 
-///@brief Method for mapping the Memgraph in-memory graph into user-based graph
-/// with iterative node indices. The graph object stores information about the node
-/// and edge mappings alongside connection information. Created graph is
-/// zero-indexed, i.e. indices start at index 0.
+///@brief Maps the Memgraph in-memory graph to a user-defined graph view.
+/// Supports building unweighted and weighted graph views, as well as views of
+/// subgraphs with given nodes and relationships.
+/// Node/relationship IDs are zero-indexed.
+/// The graph view holds connection information and local<->Memgraph ID mapping.
 ///
-///@param graph Memgraph graph
-///@param result Memgraph result object
-///@param memory Memgraph storage object
-///@param graph_type Graph directedness
-///@return mg_graph::Graph
+///@param graph -- Memgraph graph
+///@param result -- Memgraph result object
+///@param memory -- Memgraph storage object
+///@param graph_type -- graph directedness
+///@param weighted -- graph weightedness
+///@param weight_property -- relationship property containing weight
+///@param default_weight -- default relationship weight if property not set
+///@param subgraph -- if set, limit the graph view to subgraph
+///@param subgraph_nodes -- nodes selected for subgraph
+///@param subgraph_relationships -- relationships selected for subgraph
 ///
+///@return mg_graph::Graph -- graph view
 template <typename TSize = std::uint64_t>
-std::unique_ptr<mg_graph::Graph<TSize>> GetGraphView(mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory,
-                                                     const mg_graph::GraphType graph_type) {
+std::unique_ptr<mg_graph::Graph<TSize>> GraphViewGetter(
+    mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory, const mg_graph::GraphType graph_type,
+    const bool weighted = false, const char *weight_property = "weight", double default_weight = 1.0,
+    const bool subgraph = false, mgp_list *subgraph_nodes = nullptr, mgp_list *subgraph_relationships = nullptr) {
   auto graph = std::make_unique<mg_graph::Graph<TSize>>();
 
+  std::unordered_set<std::uint64_t> subgraph_node_ids, subgraph_edge_ids;
+  if (subgraph) {
+    for (std::size_t i = 0; i < mgp::list_size(subgraph_nodes); i++) {
+      subgraph_node_ids.insert(mgp::vertex_get_id(mgp::value_get_vertex(mgp::list_at(subgraph_nodes, i))).as_int);
+    }
```

Comment:
Why don't we save here reference to `vertex`, and later use that reference, instead of checking inside `vertices_it` whether there exists a node with such `id`?

Reply:
Understanding your viewpoint, I would like to raise the following concern: when the user calls sth. like `MATCH (a:Node)-[e]->(b:Node) WHERE {criterion} WITH collect(a) AS nodes, collect(e) AS edges ...`,  `nodes` will contain duplicates, e.g.
* if edges `3->5` and `4->5` are `MATCH`ed, node `5` is duplicated
* undirected `MATCH`ing like `MATCH (a:Node)-[e]-(b:Node)` doubles *everything*
As for storing `mgp` types in sets, `set<mgp_vertex>` and `set<mgp_edge>` cause errors as `mgp_vertex` and `mgp_edge` are incomplete types. Extending the types calls for changing the MGP API; I’d already consulted Benjo about that (with iteration over mgp containers) and he was against because the API should be as C-like as possible
Doing `set<mgp_vertex*>` & `set<mgp_edge*>` doesn’t work either because duplicates have different hashes

With this in mind, I‘m keeping this part as-is for now.

Reply:
I’ve succeeded in resolving this by adding duplicate checks to the `CreateNode` and `CreateEdge` methods of graphview.